### PR TITLE
feat: switch global ratelimit config provider to xds-grpc-sotw based server

### LINKED ISSUE #1254: ratelimit: switch to XdsGrpcSotwProvider
For now, we use [FileProvider](https://github.com/envoyproxy/ratelimit/blob/main/src/provider/file_provider.go#L16) in RLS

After https://github.com/envoyproxy/ratelimit/pull/373, [XdsGrpcSotwProvider](https://github.com/envoyproxy/ratelimit/blob/main/src/provider/xds_grpc_sotw_provider.go#L31) can be used to speed config propagation. 

cc @arkodg

### PR DESCRIPTION
**Pull request by shawnh2**
**What type of PR is this?**

feat(global-ratelimit): switch config provider to xds-grpc-sotw based

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #1254

### CODE REVIEW

File: internal/globalratelimit/runner/runner.go
Comment (shawnh2) [inline internal/globalratelimit/runner/runner.go:62]: do we need to enable a TLS connection with the xDS confi server?

File: internal/infrastructure/kubernetes/ratelimit/resource.go
Comment (shawnh2) [inline internal/infrastructure/kubernetes/ratelimit/resource.go:47]: I haven't expose this port in envoy-gateway service yet. see if is ok to open this port for xds config server? @arkodg

File: internal/infrastructure/kubernetes/ratelimit/resource.go
Comment (shawnh2) [inline internal/infrastructure/kubernetes/ratelimit/resource.go:140]: for now it passes the test using cases in https://gateway.envoyproxy.io/v0.4.0/user/rate-limit.html, but still need to test `ConfigType: FILE` mode out, in case someone is tampering with this variable

File: internal/globalratelimit/runner/runner.go
Comment (arkodg) [reply internal/globalratelimit/runner/runner.go:62]: yes, can be done as a follow up

File: internal/infrastructure/kubernetes/ratelimit/resource.go
Comment (arkodg) [reply internal/infrastructure/kubernetes/ratelimit/resource.go:47]: sure since 
```
DefaultXdsServerPort = 18000
```

**Comment by arkodg** [comment]
that was quick @shawnh2 ! will take another look when its out of draft

File: internal/infrastructure/kubernetes/ratelimit/resource.go
Comment (shawnh2) [reply internal/infrastructure/kubernetes/ratelimit/resource.go:47]: okay, sure

File: internal/infrastructure/kubernetes/ratelimit/resource.go
Comment (shawnh2) [reply internal/infrastructure/kubernetes/ratelimit/resource.go:140]: if so, do we need to prevent someone is changing the `ConfigType` to `FILE`, maybe add some validations?

**Review by zirain — COMMENTED** [review]
LGTM, defer to @arkodg

File: internal/globalratelimit/runner/runner.go
Comment (Xunzhuo) [inline internal/globalratelimit/runner/runner.go:88]:
```
if err := r.grpc.Serve(l); err != nil {}
```

File: internal/globalratelimit/runner/runner.go
Comment (Xunzhuo) [inline internal/globalratelimit/runner/runner.go:105]:
```
if err := r.addNewSnapshot(ctx, nil) ; err != nil {
}
```

File: internal/globalratelimit/runner/runner.go
Comment (Xunzhuo) [inline internal/globalratelimit/runner/runner.go:113]: From line 113 to 131, can you do some opts on them ? Too many if else and make it less readable.

File: internal/infrastructure/kubernetes/ratelimit/resource.go
Comment (Xunzhuo) [inline internal/infrastructure/kubernetes/ratelimit/resource.go:204]: Do we need to check if `XdsGrpcSotwConfigServerPort` is a valid port ?

**Review by Xunzhuo — COMMENTED** [review]
Some nits, generally LGTM.

File: internal/globalratelimit/runner/runner.go
Comment (shawnh2) [reply internal/globalratelimit/runner/runner.go:113]: wrapped config cache updating logic into a method, should be better now.

File: internal/infrastructure/kubernetes/ratelimit/resource.go
Comment (shawnh2) [reply internal/infrastructure/kubernetes/ratelimit/resource.go:204]: right now the `DefaultXdsServerPort` does not have a check either.

https://github.com/envoyproxy/gateway/blob/89b8b398b7c2bf2caa8452f4c72a6228feb2db30/internal/xds/bootstrap/bootstrap.go#L29-L30

maybe we can make these two ports configureable, and done as a follow-up?

**Comment by zirain** [comment]
@shawnh2 I think this will speed up `e2e` test, can you try to change following configuration, let's see what will happen.
https://github.com/envoyproxy/gateway/blob/89b8b398b7c2bf2caa8452f4c72a6228feb2db30/test/e2e/tests/ratelimit.go#L38

File: internal/infrastructure/kubernetes/ratelimit/resource.go
Comment (arkodg) [reply internal/infrastructure/kubernetes/ratelimit/resource.go:200]: maybe the user want want to override to `FILE` manually, its fine to ignore that case imo

File: internal/globalratelimit/runner/runner.go
Comment (arkodg) [inline internal/globalratelimit/runner/runner.go:149]: curious why is the service config needed now that we are using xds ?

**Comment by arkodg** [comment]
PR looks good ! just added some minor comments

File: internal/globalratelimit/runner/runner.go
Comment (shawnh2) [reply internal/globalratelimit/runner/runner.go:149]: because for `ratelimitInfra`, we only use it to update the configMap `envoy-ratelimit`:

https://github.com/envoyproxy/gateway/blob/a9a85281f03cada583f65dca61c111e7c36e10d9/internal/infrastructure/kubernetes/ratelimit/resource_provider.go#L48-L50

that configMap `envoy-ratelimit` is used when we run ratelimit provider in FILE mode.

like the discuss above, someone may want to override the XDS mode to FILE manually. so if people does that, we can still make sure that ratelimit provider is working.

**Comment by shawnh2** [comment]
> @shawnh2 I think this will speed up ratelimit `e2e` test, can you try to change following configuration, let's see what will happen.
> 
> https://github.com/envoyproxy/gateway/blob/89b8b398b7c2bf2caa8452f4c72a6228feb2db30/test/e2e/tests/ratelimit.go#L38

@zirain hi, I shrink the sleep time down to 30s now, to see how it goes.

i notice that there're still some TODO in ratelimit e2e test, i can help with that as follow-up.
https://github.com/envoyproxy/gateway/blob/a9a85281f03cada583f65dca61c111e7c36e10d9/test/e2e/tests/ratelimit.go#L36

File: internal/globalratelimit/runner/runner.go
Comment (arkodg) [reply internal/globalratelimit/runner/runner.go:149]: imo we should remove all the configmap logic, its more maintenance burden  for this project to support multiple config methods. afaik there is no use case to support both, users would like to configure ratelimiting and are not concerned with how ratelimiting is being implemented

Comment (shawnh2) [reply internal/globalratelimit/runner/runner.go:149]: agree, I've removed the `FILE` mode.

File: internal/ir/ratelimit_infra.go
Comment (arkodg) [inline internal/ir/ratelimit_infra.go:12]: can you elaborate why this is needed ?

File: internal/infrastructure/kubernetes/ratelimit/resource_provider.go
Comment (arkodg) [inline internal/infrastructure/kubernetes/ratelimit/resource_provider.go:44]: can we delete the entire ConfigMap ? dont think its needed anymore

File: internal/ir/ratelimit_infra.go
Comment (shawnh2) [reply internal/ir/ratelimit_infra.go:12]: since we don't use RateLimitServiceConfig anymore, I rm it.

https://github.com/envoyproxy/gateway/blob/334422166f208a03440fcec1534c8fe4a2aab3ef/internal/ir/ratelimit_infra.go#L18-L23

but we do need to keep RateLimitInfra, 

https://github.com/envoyproxy/gateway/blob/334422166f208a03440fcec1534c8fe4a2aab3ef/internal/ir/ratelimit_infra.go#L10-L13

so when we Subscribe the changes from xdsIR channel, like someone enable the global ratelimit:

https://github.com/envoyproxy/gateway/blob/334422166f208a03440fcec1534c8fe4a2aab3ef/internal/globalratelimit/runner/runner.go#L46

the RateLimitInfraIR should be informed, but how can we tell whether is delete/create? so I use `ServiceNames []string`, it's just derived from `[]*RateLimitServiceConfig` but only contains `name`.

```go
if rlIR == nil {
	r.RateLimitInfraIR.Delete(r.Name())
} else {
	r.RateLimitInfraIR.Store(r.Name(), rlIR)
}
```

and infra Subscribe to the RateLimitInfraIR, so it can create/delete the service and deployment of ratelimit.

https://github.com/envoyproxy/gateway/blob/334422166f208a03440fcec1534c8fe4a2aab3ef/internal/infrastructure/runner/runner.go#L76-L79

Comment (arkodg) [reply internal/ir/ratelimit_infra.go:12]: ah thanks for flagging this, I think more deletions are needed :)
* we need to get rid of https://github.com/envoyproxy/gateway/blob/main/internal/ir/ratelimit_infra.go since there is no useful info being passed around from the rl-service to the infra-manager
* we need to get rid of the message https://github.com/envoyproxy/gateway/blob/554176b8cbbc5ffa627dc7109e1b3d2b59474e99/internal/message/types.go#L72 since its no longer useful
* the subscriber in the infra-manager needs to be deleted
https://github.com/envoyproxy/gateway/blob/554176b8cbbc5ffa627dc7109e1b3d2b59474e99/internal/infrastructure/runner/runner.go#L76 since its no longer useful
* the infra manager should deploy a RL service based on whether rate limit is configured or not
https://github.com/envoyproxy/gateway/blob/554176b8cbbc5ffa627dc7109e1b3d2b59474e99/internal/infrastructure/runner/runner.go#L48
it doesnt need to be dynamic

All these pieces can be handled in this PR or a follow up PR is fine, dont want to block this PR, since its tremendous improves the time for ratelimit configuration

**Review by arkodg — APPROVED** [review]

**Review by qicz — APPROVED** [review]
great job

**Review by zirain — APPROVED** [review]

**Review by zirain — APPROVED** [review]

File: internal/ir/ratelimit_infra.go
Comment (shawnh2) [reply internal/ir/ratelimit_infra.go:12]: sure, the more deletions will be done as follow-up

File: internal/globalratelimit/runner/runner.go
Comment (arkodg) [reply internal/globalratelimit/runner/runner.go:62]: raised https://github.com/envoyproxy/gateway/issues/1454 to track it

File: internal/ir/ratelimit_infra.go
Comment (arkodg) [reply internal/ir/ratelimit_infra.go:12]: resolved with https://github.com/envoyproxy/gateway/pull/1451

**Pull request merged**
